### PR TITLE
Build in docker

### PR DESCRIPTION
**Pull request by a-magdy**
Add some files to enhance the docker experience, they of course need to be renamed and moved

I am just sharing them here, getting your opinion and will update it later

I make this to avoid having to build the app on mac: #60 

Things I added:
- Add `.dockerignore` file
- Add docker/build-base to create a base image to build the app (with the requirements) -could be moved to its own repo with Docker hub automated builds-
- Add docker-build (uses the docker/build-base output image) & docker-build-from-scratch dockerfiles to build the app in it (maybe we can modify it to different needs
- Add a script to build the app in docker, then copy the output to `docker-build` folder on the host
- Add 2 versions of the a multi-stage builds to demonstrate how we can build different types (with optimized final images) completely automated in docker (maybe again use Docker hub automated builds)

TODOs:
- One of the things I want to do it contribute more to the library, and I think providing a universal way to build and run the tool will be very useful
- Create a way to run monstache from the output of source code (to test changes quickly), maybe create a monstache-dev docker image, which we send the code as a volume, use a watcher and rebuild the tool, and use the result as the binary of the docker image
- Create a structure to run tests without depending on the localhost ES & MongoDB, I for ex don't have both locally, I run them in docker containers only, another use would be to use something like Gitlab CI to run tests in docker automatically with every commit (Or CircleCI or Travis in GitHub)

How to Use:
1. Build the `builder-base` image
  - Go to `docker/builder-base`, then the script `build.sh` to build an image which will be used to build the app later
2. Test different docker files in the root directory, by running `docker build . -f <docker_file_name> -t <target_image_name>
  - Ex: `docker build . -f Dockerfile-build-multi-stage -t rwynn/monstache:<version_number>`

### CODE REVIEW

**Comment by rwynn** [comment]
Thanks for submitting this.  It may take me a little while to parse it all but it looks really good from what I've understood so far.  I've been wanting to automate things a bit more because it's a real pain to make each release especially with the 2 different versions to support different Elasticsearch APIs.  

My docker-fu is pretty minimal at this point.  I mostly do the development in vagrant boxes which is like so 00s.  Actually, it's not too bad since it's one command to have the MongoDB and Elasticsearch servers running.

Automated builds and tests would be really nice.  I've been meaning to look into Gitlab, but maybe Travis would be better since this is already on Github?  I've only used Travis once before, but I remember it being pretty easy to set up.

**Comment by a-magdy** [comment]
Sure, I'd like to help you with creating a new pipeline if you allow me (with automating docker hub builds and GitLab CI), I'd done this before

Maybe we could write down the requirements for automating monstache completely, and work on it piece by piece

**Comment by a-magdy** [comment]
I want to move forward with this as I think the app is getting more attraction, and we'll to be able to build it faster and test more features

I suggest moving extra (testing) dockerfiles somewhere else, and leave only one of them, that build the app from scratch `dockerfile-build-multi-stage-from-scratch`, and just name it `Dockerfile`

I just want to move this forward with the least work required from you :)

**Comment by a-magdy** [comment]
I am with you in any decision that helps the most

I have a couple of concerns but I think we can find solutions for them.

The only thing that I am not sure how we are going to fix, is CI, we'll need to build images from CI with changes in the monstache repo itself, how are we going to do that if the scripts to build are in another repo?

Maybe we could put in this repo the minimal requirements to build and test monstache using docker, and create another repo with extra features!

Also most project I see online, have docker files in the source code, as the build scripts change with the app itself, so if you want to be able to build any commit at anytime, it is easier if the build files are part of the changes

And please note, that all these docker files/scripts I wrote, are only to show what we can do with them, but we'll choose one of them to build, and the rest are not going to be in the root directory

If I choose, I'd say remove everything, and just go for the `dockerfile-build-multi-stage-from-scratch` as the default one, name it `Dockerfile` as it can build a docker image with minimal size, with no requirements

But other scripts can help extract different build targets, or make use of pre-built image (to speed up build time) .. not required, but nice to have

Summary:
- I think having them in the source code would help a lot with CI, testing changes faster
- We'll reduce the number of files to the minimum required files, and move the rest to `scripts` or `.bin` folder

**Comment by rwynn** [comment]
I think you’re right about keeping it in the same repo. I just read https://docs.docker.com/docker-hub/builds/ and I think it would work if we had 2 sub folders under the root with a Dockerfile in each. One would be for a regular build and another for a CGO=0 build.  There would be nothing to uncomment or change in the Dockerfile. This folder structure would then be copied to the rel3 branch and adjusted as necessary. Then I could create 4 automated docker builds by referencing the 2 docker files in each of the 2 branches. Like you said maybe these should be based off of multi-stage-from-scratch.

**Comment by a-magdy** [comment]
I like this plan

I'll update the files this weekend according to your plan, we'll have a Dockerfile in root directly without CGO=0, and then I'll work on the other on in another folder (which image is it based on btw?)

And for Docker hub builds, tagging images might be a little bit tricky

When you create an automated build, you specify the tag name of this build, something like this:

skaermbillede 2018-05-05 kl 19 12 19

So I was thinking that it might be easier if the ones in master are called `nightly` build or simply `latest`, and when you have a release, you can build it from another branch or another tag, so the config would look like this:

skaermbillede 2018-05-05 kl 19 15 33

**Comment by rwynn** [comment]
Hi @a-magdy, I've been looking into integrating the work you've done so far.  I just pushed a feature branch to share my ideas so far.  

Basically, I think the work you did to incorporate musl-dev removes the need to have 2 separate builds for each branch.  I was able to use the `builder-release` folder to build an image that works fine with go plugins.  So I don't think we actually need 2 builds anymore which is great.  

My next move was going to be to investigate how to make an automated docker build that uses the Dockerfile in builder-release and uses the details of this thread to pass in the build-args/customize via a build hook.

https://github.com/docker/hub-feedback/issues/508

Basically, I was going to see if there was a way to make the build hook look like `build.sh` and if not use a simple build hook with things stubbed out.  I think this automated build should be labeled `latest` like you mentioned.  

However, since a binary for the 3.X and 4.X versions of the code maybe I need to separate monstache into `monstache` (tracks master) and `monstache3` (tracks rel3) so both would be able to have their own `latest` label.

**Comment by a-magdy** [comment]
Great, I think we can do this on multiple steps:

- I'll clean up the branch now with the changes we agreed on (with 1 Dockerfile in root directory initially)
- Create CI tests using CircleCI or TravisCI
- Create another repo to build a builder docker image: https://github.com/Quadric/monstache/blob/build-in-docker/docker/builder-base/Dockerfile-builder-base which will speed up CI run time, and reduce build minutes required in 1 month, as we only have 1500mins/month
- Enhance the process with multiple build processes / multiple Docker Hub image for different releases (v3 & v4)

I'll work now of cleaning up the branch, and can you please create another repo for the builder image and give me access to it, we could call it `monstache-builder-base` or `monstache-builder` or you can choose a better name, it'll basically container Dockerfile that builds an image based on golang:alpine with all required dependencies

**Comment by a-magdy** [comment]
I removed all unneeded files, and now we have only 1 Dockerfile in the root directory to build an image

We can easily create another version with CGO=0 if you want
